AlphaWallet/TokenScriptClient/Models/EventsDataStore.swift

```diff
@@ -63,13 +55,37 @@ class EventsDataStore: EventsDataStoreProtocol {
         }
     }
 
-    func add(events: [EventInstance], forTokenContract contract: AlphaWallet.Address) {
-        guard !events.isEmpty else { return }
-        try! realm.write {
-            for each in events {
-                realm.add(each, update: .all)
+    func getMatchingEventsSortedByBlockNumber(forContract contract: AlphaWallet.Address, tokenContract: AlphaWallet.Address, server: RPCServer, eventName: String) -> Promise<[EventInstanceValue]> {
+        return Promise { seal in
+            let events = Array(realm.threadSafe.objects(EventInstance.self)
+                .filter("contract = '\(contract.eip55String)'")
+                .filter("tokenContract = '\(tokenContract.eip55String)'")
+                .filter("chainId = \(server.chainID)")
+                .filter("eventName = '\(eventName)'")
+                .sorted(byKeyPath: "blockNumber"))
+                .map { EventInstanceValue(event: $0) }
+
+            seal.fulfill(events)
+        }
+    }
+
+    func add(events: [EventInstanceValue], forTokenContract contract: AlphaWallet.Address) -> Promise<Void> {
+        if events.isEmpty {
+            return .value(())
+        }
+
+        return Promise { seal in
+            do {
+                let realm = self.realm.threadSafe
+                try realm.write {
+                    let eventsToSave = events.map { EventInstance(value: $0) }
```

Comment:
So `EventInstance(value: $0)` uses KVC? So it might be possible for it to break when the properties no longer match. Can you see if you can write a test class that helps to prevent this error in the future? It doesn't have to be a test case that checks at runtime (e.g. `ServersCoordinatorTests.testServerListIsComplete`), it could be be something that just fails to build if you just list out the properties for both and add a comment that the programmer should make sure they are the same. Eg.

//a comment here to describe what this test case helps to guard against
let v = EventInstanceValue(contract: ...)
let _ = v.chainId
let _ = v.blockNumber
//other properties
//repeat for EventInstance with the same list of properties as above

Reply:
it seems like 

```
@objc dynamic var json: String = "{}" {
        didSet {
            _data = EventActivity.convertJsonToDictionary(json)
        }
    }
```
will never get called, because it sets only in `init` method.